docs/design/simplified_federation_deployment.md

```diff
@@ -0,0 +1,387 @@
+
+**1. Remove the cluster resource from federation**
+
+* Remove the cluster API resource corresponding to `<cluster-context-
+name>` from federation.
+
+**2. Remove the credentials secret**
+
+* Remove the kubeconfig secret that we created during the registration
+  process from the bootstrap cluster.
+
+**Note:** `--bootstrap-cluster` flag is only required because federation
+still doesn't support selecting specific clusters in federated secrets.
+It copies them to all the underlying clusters. It is unnecessary to copy
+the credentials in clusters other than the bootstrap cluster. And since
+these are secrets that include cluster credentials, we want to err on
+the side of being too conservative. This flag can be deprecated once
+federated secrets gain cluster selection support.
+
+## Test Plan
+
+#### Federation Control Plane Bootstrap
+
+End-to-end testing of `kubeadm` functionality is somewhat tricky because
+`kubeadm init` and hence `kubeadm init federation` prints out a token to
+STDOUT that must be used in all the `kubeadm join` invocations on the
+non-master Kubernetes nodes. Given that, here is an overview of how we
+plan to test the federation bootstrap functionality.
+
+**1. kubernetes-anywhere for resource provisioning**
+
+* We use
+  [kubernetes-anywhere](https://github.com/kubernetes/kubernetes-anywhere)
+  to provision cloud resources, i.e. VMs, Networking etc.
+
+**2. Provide a customized VM startup script**
+
+* We provide a customized `configure-vm.sh` script to run during VM
+  startup.
+* On the master VM, this script runs `kubeadm init federation` which
+  prints out a token.
+* The token is written to an object store (GCS on GCP and S3 on AWS).
+* On the non-master nodes, `configure-vm.sh` script polls for the token
+  object in the pre-configured storage bucket.
+* Once that appears, it uses the token to run `kubeadm join <token>`.
+
+**3. Verify that the federation control plane is up and running**
+
+* We verify that the federation control plane pods are up and are in
+  "Running" state.
+* We verify that the request to a discovery endpoint on the federation
+  API server returns a valid response.
+
+#### Cluster Registration/Deregistration
+
+This should be as simple as running `kubectl register <cluster-context>`
+and verifying that the registered cluster reports "Ready" status.
+
+Running `kubectl get clusters` should not list the unregistered cluster.
+
+
+## Timeline
+
+Both the federation control plane deployment functionality via `kubeadm`
+and the cluster registration/deregistration functionality via `kubectl`
+are planned to be implemented in the v1.5 time frame and are considered
+a P0 feature for v1.5.
+
+
+## Maturity Level
+
+Both functionalities will be labeled "alpha" in v1.5.
+
+
+<!-- BEGIN MUNGE: GENERATED_ANALYTICS -->
+[![Analytics](https://kubernetes-site.appspot.com/UA-36037335-10/GitHub/docs/design/simplified_federation_deployment.md?pixel)]()
+<!-- END MUNGE: GENERATED_ANALYTICS -->
```

Comment:
1 - Is there some design reason, the register and de-register of clusters have to be kept in a separate step or using a separate tool(apart from one obvious fact that the underlying cluster should have one healthy node join the master first, to have the control plane components up and running).
    I mean, kubeadm from the name, it appears to be a tool to be used by admins 
    My end use view of of federation is as follows:
    - end user can be an admin user (who sets up and manages clusters), or a normal user (who accesses clusters to create objects, rc, pods, services etc)
    - as a cloud admin, 
       I have multiple clusters which i want to federate
       I collate info about all clusters (endpoints, credentials, etc) at one location
       Then I say federate these clusters (this to me is an admin action)
       Once the above is complete, I distribute the federation api end point (and credentials, if there are) to rest of the users.

    A suggestive thought would be to have more admin level actions (for example deploying control plane, registering clusters) can be kept as part of kubeadm and more general puspose ones (creating/deleting all other objects, status queries, etc)as part of kubectl. But this is just a suggestion.. :)


2 - I cannot visualize (as of now) how are you planning to bring out the point of deploying federation control plane over multiple clusters (for HA or related stuff). I think to accommodate the multi cluster federation control plane deployment also, it might make sense to separate the overall process into 3 properly demarcated steps.
   step 1 - k8s cluster deployment (which need not be part of federation control plane deployment at all, only a prerequisite, for example i have cluster A and cluster B available on which I want to deploy my federation control plane). In most cases, I would also have a list of existing clusters which I want to federate (offcourse, I can have more clusters join this federation, or leave it, later)
   step 2 - federate the clusters (which can include bring up control plane, somewhere, and join all clusters together)
   step 3 - actual end user usage (who might be oblivious to this fact also that they are talking to federation, to them its a big k8s cluster)